### DISCUSSION TITLE
How to add extra edges to graph data?

### DISCUSSION THREAD

**Discussion by songsong0425** [pyg-team/pytorch_geometric#7489] [Q&A]
Hi, I'm trying to build a homogeneous graph for GCN.\
During the process, I'd like to expand the `Data` by adding extra edges. \
For the initial graph construction, I ran the below code:
```python
# Generate Data object by allocating edge_index and edge_type
edge_index = torch.zeros((2, len(mapped_relation_df)), dtype=torch.long)
edge_type = torch.zeros(len(mapped_relation_df), dtype=torch.long)

for i in range(len(mapped_relation_df)):
    edge_index[0,i] = mapped_relation_df.iloc[i][0]
    edge_index[1,i] = mapped_relation_df.iloc[i][2]
    edge_type[i] = mapped_relation_df.iloc[i][1]
    
homo_g = Data(edge_index=edge_index, edge_type=edge_type)
homo_g.num_nodes = len(entity_dict)
homo_g
#Data(edge_index=[2, 2159252], edge_type=[2159252], num_nodes=47824)
```

After that, I want to combine other graph data by connecting the edge index.

```python
df
#         A        B        A_1        B_1
# 0       11736    322      47824      47884
# 1       15205    14489    47825      47867
# ...	...	...	...	...	...
# 1078    8808     5342     47907      47939
# 1079    19208    1881     47921      47956
```
A and B columns are matched id in `homo_g` and `A_1` and `B_1` are new indices for the future node features.\
In other words, `edge_index[0]` contains `A` and `B`, and `edge_index[1]` contains `A_1` and `B_1`.\
As a result, I expect the data like below figure:
![ex](https://github.com/pyg-team/pytorch_geometric/assets/80442548/e46aa6a9-f89f-4ecb-bed7-3c582fb621b6)

**Comment by wsad1** [comment]
Could you clarify this please? Do you want to add edges represented by `A_1` and `B_1` to your `Data` object?

**Reply by songsong0425** [reply]
Sorry for unclear description. Yes. As you said, I want to add edges `A_1` and `B_1` to `Data (i.e., homo_g)` object. \
During hours, I tried the below code:
```python
df_mappedid = torch.tensor(np.hstack(df[['A', 'B']].values.tolist()))
df_newid = torch.tensor(np.hstack(df[['A_1', 'B_1']].values.tolist()))
df_combinedid = torch.stack((df_mappedid, df_newid ))

edge_index_combine = torch.cat([homo_g.edge_index, df_combinedid], dim=1)    
final_graph = Data(edge_index=edge_index_combine, x=features)
```

But I guess that there are other simpler approaches.

**Reply by wsad1** [reply]
I think the approach you shared is the right one.

**Reply by songsong0425** [reply]
Sorry for rate reply and thank you for your comfirmation!
